File: app-ios/VelaWallet/VelaWallet/Core/RegistryNameLookup.swift
//
//  RegistryNameLookup.swift
//  VelaWallet
//
//  The name a Vela user registered for a wallet, found by its ADDRESS — the
//  iOS transport for `vela_core::registry_lookup` (issue 191).
//
//  The v2 index cannot be asked about an address: `?walletRef=` answers 400,
//  because an address is `f(all founding keys)` and no single entry owns one.
//  `RegistryClient.queryByWalletRef` asked exactly that for two specs, and the
//  index rung of `RecipientIdentity` silently never answered. The name is
//  reached through the chain instead — the Safe's `SafeWebAuthnSharedSigner`
//  configuration → `?publicKey=` → `?unitId=` — and every rule of that walk
//  (which chain first, which unit is believed, what a malformed blob means,
//  how long a verdict may be kept) is the core's, shared with the other three
//  shells. What is left here is what only a shell can do: perform the requests
//  and keep the verdict for as long as the core says.
//

import Foundation
import VelaCore

@MainActor
final class RegistryNameLookup {
    /// How one GET against the index ended.
    enum IndexAnswer {
        case ok(String)
        /// The index said 404: an answer.
        case notFound
        /// Unreachable, timed out, any other status: nobody said anything.
        case failed
    }

    /// `vela.indexName:<address>` — web's key and shape: `{ hit: { name,
    /// publicKey } }` or `{ missAt }`.
    static let cachePrefix = "vela.indexName:"
    /// `registry_lookup::MISS_TTL_MS` — a miss is only true for now.
    static let missTTLMs: Double = 6 * 60 * 60 * 1000
    /// A lookup is a handful of rounds; this only stops a contract bug from spinning.
    private static let maxRounds = 16

    private let store: VelaStore
    /// `eth_call` on one chain: the RAW `result` hex — a bare `0x` included,
    /// which is a chain saying "no such contract here" — or `nil` when nobody
    /// answered. The distinction is the core's caching rule.
    private let ethCall: (_ chainId: Int, _ to: String, _ data: String) async -> String?
    private let indexGet: (_ path: String) async -> IndexAnswer
    /// `registryNameStep` from the core; a seam so tests can script it.
    private let step: (_ address: String, _ answersJson: String) -> String
    private let now: () -> Double

    init(
        store: VelaStore,
        ethCall: @escaping (Int, String, String) async -> String?,
        indexGet: @escaping (String) async -> IndexAnswer,
        step: @escaping (String, String) -> String = { registryNameStep(address: $0, answersJson: $1) },
        now: @escaping () -> Double = { Date().timeIntervalSince1970 * 1000 }
    ) {
        self.store = store
        self.ethCall = ethCall
        self.indexGet = indexGet
        self.step = step
        self.now = now
    }

    func name(for address: String) async -> String? {
        let key = address.lowercased()
        guard key.count == 42, key.hasPrefix("0x") else { return nil }

        let stored = store.readObject(Self.cachePrefix + key)
        if let hit = stored["hit"] as? [String: Any], let name = hit["name"] as? String, !name.isEmpty {
            return name
        }
        if let missAt = (stored["missAt"] as? NSNumber)?.doubleValue, now() - missAt < Self.missTTLMs {
            return nil
        }

        var answers: [[String: Any]] = []
        for _ in 0..<Self.maxRounds {
            guard let transcript = try? JSONSerialization.data(withJSONObject: answers),
                  let next = try? JSONSerialization.jsonObject(
                      with: Data(step(key, String(decoding: transcript, as: UTF8.self)).utf8)
                  ) as? [String: Any]
            else { return nil }

            if next["type"] as? String == "ask" {
                let requests = next["requests"] as? [[String: Any]] ?? []
                // "Together, if the shell can": the second tier is six chains.
                let round = await withTaskGroup(of: [String: Any].self) { group in
                    for request in requests {
                        group.addTask { [weak self] in await self?.perform(request) ?? [:] }
                    }
                    var collected: [[String: Any]] = []
                    for await answer in group where !answer.isEmpty { collected.append(answer) }
                    return collected
                }
                answers.append(contentsOf: round)
                continue
            }

            let found = next["found"] as? [String: Any]
            let name = (found?["name"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            switch next["remember"] as? String {
            case "forever":
                if let name {
                    store.writeObject(Self.cachePrefix + key, [
                        "hit": ["name": name, "publicKey": found?["public_key"] as? String ?? ""],
                    ])
                }
            case "briefly":
                if name == nil { store.writeObject(Self.cachePrefix + key, ["missAt": now()]) }
            default:
                break
            }
            return name
        }
        return nil
    }

    private func perform(_ request: [String: Any]) async -> [String: Any] {
        let id = request["id"] as? String ?? ""
        func answer(_ outcome: String, _ body: String? = nil) -> [String: Any] {
            ["id": id, "outcome": outcome, "body": body ?? NSNull()]
        }
        switch request["type"] as? String {
        case "eth_call":
            guard let chainId = (request["chain_id"] as? NSNumber)?.intValue,
                  let to = request["to"] as? String, let data = request["data"] as? String,
                  let result = await ethCall(chainId, to, data)
            else { return answer("failed") }
            return answer("ok", result)
        case "index_get":
            guard let path = request["path"] as? String else { return answer("failed") }
            switch await indexGet(path) {
            case .ok(let body): return answer("ok", body)
            case .notFound: return answer("not_found")
            case .failed: return answer("failed")
            }
        default:
            return answer("failed")
        }
    }
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/contacts/core/RegistryNameLookup.kt
package app.getvela.wallet.feature.contacts.core

import app.getvela.wallet.core.data.KeyValueStore
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import org.json.JSONArray
import org.json.JSONObject

/**
 * The name a Vela user registered for a wallet, found by its ADDRESS — the
 * Android transport for `vela_core::registry_lookup` (issue 191).
 *
 * The v2 index cannot be asked about an address: `?walletRef=` answers `400`,
 * because an address is `f(all founding keys)` and no single entry owns one.
 * `RegistryClient.nameForAddress` asked exactly that for two specs, and the
 * index step of [IdentityResolver] silently never answered. The name is
 * reached through the chain instead — the Safe's `SafeWebAuthnSharedSigner`
 * configuration → `?publicKey=` → `?unitId=` — and every rule of that walk
 * (which chain first, which unit is believed, what a malformed blob means, how
 * long a verdict may be kept) is the core's, shared with the other three
 * shells. What is left here is what only a shell can do: perform the requests
 * and keep the verdict for as long as the core says.
 */
class RegistryNameLookup(
    private val store: KeyValueStore,
    /**
     * `eth_call` on one chain: the raw `result` hex — INCLUDING a bare `0x`,
     * which is a chain saying "no such contract here" — or `null` when nobody
     * answered. The distinction is the core's caching rule.
     */
    private val ethCall: suspend (chainId: Int, to: String, data: String) -> String?,
    /** `GET <configured index><path>`. */
    private val indexGet: suspend (path: String) -> IndexAnswer,
    /** `registryNameStep` from the core. A seam so the JVM tests need no native library. */
    private val step: (address: String, answersJson: String) -> String,
    private val now: () -> Double = { System.currentTimeMillis().toDouble() },
) {
    sealed interface IndexAnswer {
        data class Ok(val body: String) : IndexAnswer

        /** The index said 404: an answer. */
        data object NotFound : IndexAnswer

        /** Unreachable, timed out, any other status: nobody said anything. */
        data object Failed : IndexAnswer
    }

    suspend fun nameFor(address: String): String? {
        if (!ADDRESS.matches(address)) return null
        val key = address.lowercase()
        cached(key)?.let { return it.name }

        val answers = JSONArray()
        repeat(MAX_ROUNDS) {
            val next = runCatching { JSONObject(step(key, answers.toString())) }.getOrNull() ?: return null
            if (next.optString("type") == "ask") {
                val requests = next.optJSONArray("requests") ?: return null
                val round = coroutineScope {
                    (0 until requests.length()).map { index ->
                        val request = requests.getJSONObject(index)
                        async { perform(request) }
                    }.awaitAll()
                }
                round.forEach(answers::put)
                return@repeat
            }
            // `optString` turns a JSON null into the string "null" (org.json);
            // `optJSONObject` is the one that says "absent".
            val name = next.optJSONObject("found")?.optString("name")?.takeIf { it.isNotBlank() }
            when (next.optString("remember")) {
                "forever" -> if (name != null) {
                    val publicKey = next.optJSONObject("found")?.optString("public_key").orEmpty()
                    remember(key, JSONObject().put("name", name).put("publicKey", publicKey))
                }
                "briefly" -> if (name == null) remember(key, JSONObject().put("missAt", now()))
            }
            return name
        }
        return null
    }

    private suspend fun perform(request: JSONObject): JSONObject {
        val id = request.optString("id")
        fun answer(outcome: String, body: String? = null) =
            JSONObject().put("id", id).put("outcome", outcome).put("body", body ?: JSONObject.NULL)
        return runCatching {
            when (request.optString("type")) {
                "eth_call" -> {
                    val result = ethCall(request.getInt("chain_id"), request.getString("to"), request.getString("data"))
                    if (result == null) answer("failed") else answer("ok", result)
                }
                "index_get" -> when (val got = indexGet(request.getString("path"))) {
                    is IndexAnswer.Ok -> answer("ok", got.body)
                    IndexAnswer.NotFound -> answer("not_found")
                    IndexAnswer.Failed -> answer("failed")
                }
                else -> answer("failed")
            }
        }.getOrElse { answer("failed") }
    }

    private class Cached(val name: String?)

    /** `null` = nothing usable stored; `Cached(null)` = a miss still inside its window. */
    private suspend fun cached(key: String): Cached? {
        val raw = store.read(CACHE_KEY) ?: return null
        val entry = runCatching { JSONObject(raw).optJSONObject(key) }.getOrNull() ?: return null
        entry.optString("name").takeIf { it.isNotBlank() && entry.has("name") }?.let { return Cached(it) }
        val missAt = entry.optDouble("missAt", Double.NaN)
        return if (!missAt.isNaN() && now() - missAt < MISS_TTL_MS) Cached(null) else null
    }

    private suspend fun remember(key: String, entry: JSONObject) {
        val all = runCatching { JSONObject(store.read(CACHE_KEY) ?: "{}") }.getOrElse { JSONObject() }
        store.write(CACHE_KEY, all.put(key, entry).toString())
    }

    companion object {
        /** `{ "0xlowercase": { name, publicKey } | { missAt } }` — one document, as the identity cache is. */
        const val CACHE_KEY = "vela.indexName"

        /** `registry_lookup::MISS_TTL_MS` — a miss is only true for now. */
        const val MISS_TTL_MS = 6.0 * 60.0 * 60.0 * 1000.0

        /** A lookup is a handful of rounds; this only stops a contract bug from spinning. */
        private const val MAX_ROUNDS = 16
        private val ADDRESS = Regex("^0x[0-9a-fA-F]{40}$")
    }
}
